unblob/handlers/archive/qnap/qnap.py

```diff
@@ -0,0 +1,158 @@
+import io
+from pathlib import Path
+from typing import Optional
+
+from dissect.cstruct import Instance
+from structlog import get_logger
+
+from unblob.file_utils import File, InvalidInputFormat, iterate_file
+from unblob.models import (
+    Endian,
+    Extractor,
+    Regex,
+    StructHandler,
+    StructParser,
+    ValidChunk,
+)
+
+logger = get_logger()
+
+FOOTER_LEN = 74
+secret = "QNAPNASVERSION4"
+
+C_DEFINITIONS = """
+    typedef struct qnap_header {
+        char    magic[6];
+        uint32  encrypted_len;
+        char    device_id[16];
+        char    file_version[16];
+        char    firmware_date[16];
+        char    revision[16];
+    } qnap_header_t;
+"""
+
+
+class QnapExtractor(Extractor):
+    def __init__(self):
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        outpath = outdir.joinpath(f"{inpath.name}.tgz")
+        with File.from_path(inpath) as file:
+            file.seek(-FOOTER_LEN, io.SEEK_END)
+            header = self._struct_parser.parse("qnap_header_t", file, Endian.LITTLE)
+            end_offset = file.tell()
+            filesize = end_offset
+            cryptor = Cryptor(secret)
+            with outpath.open("wb") as outfile:
+                for chunk in iterate_file(file, 0, header.encrypted_len, 1024):
+                    outfile.write(cryptor.decrypt_chunk(chunk))
+                for chunk in iterate_file(
+                    file,
+                    header.encrypted_len,
+                    filesize - 74 - header.encrypted_len,
+                    1024,
+                ):
+                    outfile.write(chunk)
+
+
+# verifier les versions
+class QnapHandler(StructHandler):
+    NAME = "qts"
+
+    PATTERNS = [
+        Regex(r"icpnas"),
+    ]
+
+    C_DEFINITIONS = C_DEFINITIONS
+    HEADER_STRUCT = "qnap_header_t"
+    EXTRACTOR = QnapExtractor()
+
+    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+        header = self.parse_header(file, endian=Endian.LITTLE)
+
+        if not self.is_valid_header(header):
+            raise InvalidInputFormat("Invalid QNAP header.")
+
+        filesize = file.tell()
+        end_offset = filesize
+        return ValidChunk(start_offset=end_offset - filesize, end_offset=filesize)
+
+    def is_valid_header(self, header: Instance) -> bool:
+        if header.encrypted_len < len(header):
+            return False
```

Comment:
Why `start_offset - encrypted_len` ? It's always gonna be negative since `start_offset` is always gonna be 0 and `encrypted_len` is a static size we get from the header field